dev/com.ibm.ws.security.kerberos.auth/src/com/ibm/ws/security/kerberos/auth/KerberosService.java

```diff
@@ -110,28 +123,48 @@ public Path getKeytab() {
      * @return A valid subject for the supplied principal
      */
     public Subject getOrCreateSubject(String principal) throws LoginException {
+        return getOrCreateSubject(principal, null);
+    }
+
+    /**
+     * Checks the Subject cache for an existing Subject matching the supplied principal.
+     * A simple LRU cache is used to store cache Subjects
+     * If a valid subject is not found in the cache, then a new Kerberos login is performed and
+     * the resulting Subject is cached.
+     *
+     * @param principal The principal to obtain a subject for
+     * @param pass      The password to be used via CallbackHandler. This will only be used if no password
+     *                      is found in the credential cache or keytab files first.
+     * @return A valid subject for the supplied principal
+     */
+    public Subject getOrCreateSubject(String principal, SerializableProtectedString pass) throws LoginException {
         KerberosPrincipal krb5Principal = new KerberosPrincipal(principal);
         Subject cachedSubject = subjectCache.get(krb5Principal);
         if (cachedSubject != null) {
             return cachedSubject;
         }
 
-        Subject createdSubject = doKerberosLogin(principal);
+        Subject createdSubject = doKerberosLogin(principal, pass);
 
         subjectCache.put(krb5Principal, createdSubject);
         return createdSubject;
     }
 
-    private Subject doKerberosLogin(String principal) throws LoginException {
+    private Subject doKerberosLogin(String principal, SerializableProtectedString pass) throws LoginException {
         Subject subject = new Subject();
         Krb5LoginModuleWrapper krb5 = new Krb5LoginModuleWrapper();
         Map<String, String> options = new HashMap<String, String>();
         Map<String, Object> sharedState = new HashMap<String, Object>();
+        CallbackHandler callback = pass == null ? null : new Krb5CallbackHandler(pass);
```

Comment:
I just wonder why we do not use WSCallbackHandlerImpl.java instead of Krb4CallbackHandler?

Reply:
I didn't know we had `WSCallbackHandlerImpl`, I could use that but it looks like we'd need to also supply a username for that callback, and here I just want to supply a password specifically for `PasswordCallback`s

Reply:
will change this to `WSCallbackHandlerImpl` in a follow-on item